### PR TITLE
Support for controlling max queue time

### PR DESCRIPTION
**Pull request by KrishnaM251**
This PR addresses #2901. Note that it is a work in progress.

**New Test**
Please take a look at `test_max_queue_length.py` and play around with the `max_wait_q_len` and the number of requests present in `test_prompts`.

**Description of Current Progress**
I added a new field to EngineArgs called `max_queue_length`. If an attempt to queue more requests than `max_queue_length` is made, then an error is thrown.

The test currently has 4 requests. If you set `max_wait_q_len` to 2, then this test works as intended. However, if `max_wait_q_len` is set to 3, then the test loops forever, with the following output: 
"Running: 0 reqs, Swapped: 0 reqs, Pending: 2 reqs"

However, when I step through the debugger, I get the following output repeated:
"Running: 1 reqs, Swapped: 0 reqs, Pending: 3 reqs"

**Desired Outcome**
- figure out why the requests are not getting moved to the running queue
- throw error 503 in OpenAI compatible server (should be quick to implement)

**Location of Changes**
- test_max_queue_length.py
  - throws error if an attempt to exceed waiting queue is made
  - based heavily off of llm_engine_example.py
- args_utils.py - add a new parameter called max_queue_length  to EngineArgs
  - 33 - EngineArgs.max_queue_length
  - 205 - parser.add_argument() max_queue_length
- scheduler.py - add param to SchedulerConfig object
  - 449 - SchedulerConfig init()
- config.py
  - 466 - SchedulerConfig verifyArgs()
  - 469 - get_max_queue_length
- llm_engine.py
  - 446 - check if exceeds queue len
- engine_args.rst
  - 110 - for parsing cli args

### CODE REVIEW

File: tests/test_max_queue_length.py
Comment (simon-mo) [inline tests/test_max_queue_length.py:1]: the next step is to move this into one of the test for llm engine and follow the pytest format we are using

**Comment by simon-mo** [comment]
Additionally, can you implement the right handling in the OpenAI server?

**Comment by njhill** [comment]
I think for this it would be more useful to use a metric like total tokens in queue rather than number of requests.

**Comment by simon-mo** [comment]
@njhill can you elaborate? I thought the number of sequence is pretty natural for operator to tune because it directly translate to number of users making such request.

**Comment by njhill** [comment]
@simon-mo I think what you would really care about is the ETA for new requests that join the end of the queue, so you can reject them if this is > 5 seconds for example.

The number of tokens may be a better proxy for this. Possibly also taking the max_tokens for each request into account and/or some heuristic for some heuristic for predicting how many tokens a request is likely to generate before stopping. Something to reflect how much of the kvcache that request would consume and for how long (of course this is further complicated by prefix sharing etc.)

For example a very long queue is no big deal if it's full of small requests since they will be processed very fast.

**Comment by ywang96** [comment]
My two cents are that this a tradeoff between how fine-grained you want the control to be versus complexity in tuning these params. This is somewhat similar to `--max-num-seqs` and `--max-num-batched-tokens`, and does it not make sense to have both?

**Pull request closed without merging**